### PR TITLE
Update env variables docs

### LINKED ISSUE #1894: Document our environment variables
We should have a table documenting the names, scopes, and uses of all our environment variables.

A couple of sections only mention them in passing (manual deployment instructions and configuration object docs).

### PR DESCRIPTION
**Pull request by infomiho**
Updates the Project and Deployment env variables docs. I've tried to deduplicate as much as possible. I've enumerate all the env variables that Wasp defines.

Closes #1894

### CODE REVIEW

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:49]: We haven't introduced earlier this concept of "general configuration env vars" though, right? Should we, or it is self evident?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:54]: Should we mention that this is it, only one? Somehow it feels like waht now, where is the rest. Maybe we can mention that this is it but you can also deifne your own. Ok maybe that is silly, I am not sure, but this somehow felt unfinished.

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:87]: And some other stuff I think, for example provide a link to the client in some automatic emails (auth)? Maybe not any more if flow is completely server-side, but I would maybe state is to that it sound slike CORS is not or might not be the only thing this is important for. I wonder if we should even explain why server needs it -> that info can easily become obsolete, maybe better to skip it? Not a biggie though.

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:131]: You are talking here about a single specific env var? That is a bit weird because below we have a whole table. Feels like you are talking about the whole table. Maybe put it all in the note for that one instead?

Comment (Martinsos) [inline web/docs/project/env-vars.md:131]:
```suggestion
You can use this env var during development to skip email verification. This is useful when you are testing the email sending functionality in development:
```
The part "this is useful when you are testing the email sending functionality in development" also confuses me -> how is this useful if I am testing the email sending functionality? I guess it is useful if I am not testing email sending functinoality? Hmmmm

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:133]: Ay yay let's make that clear?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:147]: Can we put this under Auth header?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:158]: Should we create special header for Keycloak?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:92]: Where did this go, it wasn't useful?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:107]: Aha this went away because it is now mentinoed in deployment/env-vars.md?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:211]: We'll talk -> why is it future tense?

File: web/docs/project/env-vars.md
Comment (Martinsos) [inline web/docs/project/env-vars.md:110]: What happened to this image?

File: web/docs/project/EnvVarsTable.tsx
Comment (Martinsos) [inline web/docs/project/EnvVarsTable.tsx:1]: Nice this is cool! I found it a bit weird that I have to construct EnvVar elements outside and then provide them as children, to me it sounded more natural to provide a list of EnvVar objects to the EnvVarsTable and then it would call `.map` on those and produce a bunch of EnvVar elements that show those. But ok it is quite unimportant.

**Review by Martinsos — COMMENTED** [review]
Nice stuff! Once we solve all these comments, I will take another look but more from the high above, to get a better grasp for where is what and what are the relations, now I was focusing more on the details. But I think the question is similar like in other PR: Do we do A.P B.P and then link from P to those, or do we do P.A and P.B and then link from A and B to those, where are parts of docs about Production and A and B are some features (check the other PR for more info).

File: web/docs/deployment/deployment-methods/_addExternalAuthEnvVarsReminder.md
Comment (infomiho) [reply web/docs/deployment/deployment-methods/_addExternalAuthEnvVarsReminder.md:2]: I've been using this for some time now and it works great 👍 , it's explained in Writing Guide https://github.com/wasp-lang/wasp/blob/main/web/docs/writingguide.md#linking-to-pages-in-the-docs

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:92]: It didn't feel useful, it talks about the specific Node.js package and they are not going to use it directly. I think the env vars file format is self-explanatory.

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:107]: I guess so, yes 😄 I moved things around so it made sense when reading the docs.

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:110]: It moved to the deployment env vars page :)

File: web/docs/deployment/deployment-methods/paas.md
Comment (infomiho) [reply web/docs/deployment/deployment-methods/paas.md:60]: This section is now:

> There's a Dockerfile that defines an image for building the server in the `.wasp/build` directory.

> To run the server in production, deploy this Docker image to a hosting provider and make sure the required env variables are correctly set up. Usually, you use the provider's dashboard UI or a CLI tool to set up these env variables.

> Check the [required server env variables](../env-vars.md#server-env-vars) and make sure they are set up for your server.

File: web/docs/deployment/deployment-methods/paas.md
Comment (infomiho) [reply web/docs/deployment/deployment-methods/paas.md:99]: Yes 👍

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:49]: This is where we introduce it :)

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:54]: It's following the same structure as the server env vars, I think this is understandable.

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:133]: I'll update this once we merge the env vars PR since the type depends on what we decide we want to do there. #2362

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:147]: I've moved the `Auth` stuff to `Development` and now this feels better i.e. I named it `OAuth providers`.

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:158]: I feel like it would confuse people, just keeping it under `OAuth Providers` makes it feel connected to the rest of the stuff.

File: web/docs/project/EnvVarsTable.tsx
Comment (infomiho) [reply web/docs/project/EnvVarsTable.tsx:1]: Screenshot 2024-12-11 at 15 13 10

I got inspired by ShadCN's way of composing things.

File: web/docs/deployment/deployment-methods/_addExternalAuthEnvVarsReminder.md
Comment (Martinsos) [reply web/docs/deployment/deployment-methods/_addExternalAuthEnvVarsReminder.md:2]: That is suprising to me because it sounds like it goes against theri docs. Just to confirm very explicitly: you can confirm this works if we release new versioned docs and these become numbered as let's say 0.15? Will these links, when clicked from 0.15 version of docs, correctly take you to the 0.15 version of these files, or instead incorrectly to the most current docs?

File: web/docs/project/env-vars.md
Comment (Martinsos) [reply web/docs/project/env-vars.md:51]: Hm but why did I ask this then? I believe I asked it because you don't really explain it at any moment, you just go for it and start listing them. That is when I was reading it, I was thinking "oh ok general configuration, what is that" and then bam, vars immediately.
Which is why I asked above, should we explain what it is or you think it is self evident - when somebody sees "general configruation", they understand what these env vars are?

File: web/docs/project/env-vars.md
Comment (Martinsos) [reply web/docs/project/env-vars.md]: Did this ever happen (I don't see a rocket :rocket: )?

File: web/docs/project/env-vars.md
Comment (Martinsos) [reply web/docs/project/env-vars.md]: Ok, as long as you don't forget about it!

File: web/docs/project/env-vars.md
Comment (Martinsos) [reply web/docs/project/env-vars.md:92]: They might need to understand how to use it though, to for example pass JSON values and similar, we had those situations, how to quote stuff. So I thoguht it is good to introduce the concept, since we are using it as na important part of Wasp, so it is clear this is not something we made up, but instead there are more resources behind it that they can look for. Drop it if you really want, but to me it seems useful to keep it.

File: web/docs/project/EnvVarsTable.tsx
Comment (Martinsos) [reply web/docs/project/EnvVarsTable.tsx:1]: Right, but they can combine multiple things probably. You instead have a single item that can go in. So no composability, but you opened space for throwing in random elements that don't make sense.

**Review by Martinsos — COMMENTED** [review]
I did another round of comments. Once we complete those, I will as a final step take a higher look from above, probably run the docs and see how they read now.
As a preparation for that, what do you say about what I asked here https://github.com/wasp-lang/wasp/pull/2375#pullrequestreview-2469721287 ? What is your view on the current structure, how do you feel about it?

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md]: Yep, it happened, sorry 🙋‍♂️

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:92]: I get what you mean, if people know what's underneath they can Google better. 

The env variables are first parsed by Haskell and then outputted to the `.env` files in the server and the client apps. If we wanted to provide full information, we should first mention Haskell's [dotenv](https://hackage.haskell.org/package/dotenv) package and then our logic for outputting the vars i.e. our `envVarsToDotEnvContent` function. And then we can mention that the variables are ingested by Node.js with the JS [dotenv](https://www.npmjs.com/package/dotenv) package.

I dislike giving this many details to users because 99% of them don't really need them and those with issues will most likely ask about it in our Discord.

File: web/docs/project/EnvVarsTable.tsx
Comment (infomiho) [reply web/docs/project/EnvVarsTable.tsx:1]: Compare
Screenshot 2024-12-16 at 15 55 03
vs.
Screenshot 2024-12-16 at 15 55 11

One of them is more readable to me, so I'd like to keep the current API. It's our internal component for docs, no need to make it so strict IMHO.

File: web/docs/project/env-vars.md
Comment (Martinsos) [reply web/docs/project/env-vars.md:51]: I wrote this because when I was reading it I was confused for a bit, so I didn't make it up. Am I representative of a typical user, hard to say. Which is why I asked how you feel about it.

If I would write sometihng, I wouldn't write what you said because it is a bit much, especially the last part with "hence". I might write someting like "Env variables used by Wasp to configure general aspects of your app, not tied to one specific feature.".
But again, is that useful or not -> not sure. I believe I provided my perspective well enough here, I will leave the final decision to you.

File: web/docs/project/EnvVarsTable.tsx
Comment (Martinsos) [reply web/docs/project/EnvVarsTable.tsx:1]: It is more readable because you used different vars! The second example has less vars, the first and second, and third and forth, have equal length of name, and they all have much shorter notes. So there is less content, it is better aligned, and on top of all that it allowed you to zoom in so that also looks better.
Finally, your editor seems to do a better job with syntax coloring in the second example, but I don't see why it couldn't do coloring well also in the first example, I just tried this in my emacs and it does the coloring properly. Maybe that `<span>` in that note causes you issues.

Take a look at what it looks in my emacs, when I use the same env vars:

image

image

Those look exactly the same regarding readibility to me. If anything, you could argue that the second one has a bit less noise due to less repetitive HTML tags, although I don't think it matters.

Also, while doing the comparison, you ignored my argument about the second approach (one with children) allowing for elements that you might not expect, while the first one doesn't allow that.

Keep the API as you like it, as I said in my initial comment, this is quite unimportant, but still I think this discussion is useful as a learning opportunity and didn't want it to end with wrong arguments.

**Review by Martinsos — COMMENTED** [review]
Responded again.

File: web/docs/project/EnvVarsTable.tsx
Comment (infomiho) [reply web/docs/project/EnvVarsTable.tsx:1]: Updated the API 👍

File: web/docs/project/env-vars.md
Comment (infomiho) [reply web/docs/project/env-vars.md:51]: Added `These are some general env variables used for various Wasp features:` 👍

**Review by Martinsos — APPROVED** [review]

**Pull request merged**
